GUI v9.0

Wang-yuanlu opened the pull request:
Hi Mr.Tan.I  added scrollbars to the cgraPannel, aligned the xbarButton and fuCheckButton in the parampannel module, and made some slight adjustments to the GUI to make it more useful."

tancheng commented on launchUI.py:
```diff
@@ -669,7 +669,7 @@ def leave(event):
     widgets["fuConfigPannel"].config(text='Tile '+str(ID)+' functional units')
     widgets["xbarConfigPannel"].config(text='Tile '+str(ID)+' crossbar outgoing links')
-    widgets["xbarConfigPannel"].grid(columnspan=4, row=7, column=0,rowspan=2)
+    widgets["xbarConfigPannel"].grid(columnspan=4, row=7, column=0,rowspan=2,sticky="nsew")
```
Can you add a comment here about the meaning of "nsew"?

Wang-yuanlu replied:
Yes. In this sentence "nsew" means that each xbar component will fill the corresponding grid space. Similar usage is seen in the subsequent code.

tancheng commented on launchUI.py:
```diff
@@ -696,7 +696,7 @@ def clickSPM():
     spmConfigPannel.config(text='DataSPM outgoing links')
-    spmConfigPannel.grid(row=7,column=0,rowspan=2,columnspan=4)
+    spmConfigPannel.grid(row=7,column=0,rowspan=2,columnspan=4,sticky="nsew")
 
```
Can add space after ","?

tancheng commented on launchUI.py:
```diff
@@ -1449,9 +1447,9 @@ def create_cgra_pannel(master, rows, columns):
-
+    spmLabel = tkinter.Button(canvas, text = "Data\nSPM", fg = 'black', bg = 'gray', relief = 'raised', bd = BORDER, command = clickSPM)
+    #button.place(height=memHeight, width=MEM_WIDTH, x = 25, y = 35)
+    canvas.create_window(baseX+BORDER, BORDER, window=spmLabel, height=GRID_HEIGHT, width=MEM_WIDTH, anchor="nw")
```
Can you add a comment explaining the `anchor`?

tancheng replied:
Didn't see how this is resolved? Once you resolve a comment, can you please reply in this dialogue window (instead of just closing it), so it is more clear to the reviewer?

Wang-yuanlu replied:
Sorry, I forgot this one. Here I use anchor="nw" to express that the data memory will be placed in the upper left corner of the cgrapannel. This way, it will be adaptive to changes in window size.

tancheng commented on launchUI.py:
```diff
@@ -1616,19 +1619,19 @@ def create_param_pannel(master):
         spmConfigPannel.rowconfigure(i,weight=1)
     
-    spmConfigPannel.columnconfigure(0,weight=2)
+    spmConfigPannel.columnconfigure(0,weight=1)
```
Can you add a comment here about what are the `columnconfigure` used for?

Wang-yuanlu replied:
Yes. Here the columnconfigure means that the spmConfigPannel will be evenly divided into three columns, and each column will fill the corresponding grid space.

tancheng commented on launchUI.py:
```diff
@@ -1684,6 +1687,8 @@ def create_param_pannel(master):
         else:
             spmDisabledListbox.insert(0, port)
 
+
```
Remove these two lines

tancheng reviewed (commented):
I remember you and MMJ mentioned that the compilation will also be fixed in this PR?

Wang-yuanlu commented:
Hi Mr.Tan. I added comments to the code and modified the corresponding statements. I reviewed the execution method of launchUI.py in the ReadME file and I found that as long as I follow the steps in the original ReadME file, cd build, then python /WORK_REPO/CGRA-Flow/launchUI.py, the compile app function can be used normally.
Have a nice day!

tancheng commented on launchUI.py:
```diff
@@ -669,7 +666,7 @@ def leave(event):
     widgets["fuConfigPannel"].config(text='Tile '+str(ID)+' functional units')
     widgets["xbarConfigPannel"].config(text='Tile '+str(ID)+' crossbar outgoing links')
-    widgets["xbarConfigPannel"].grid(columnspan=4, row=7, column=0,rowspan=2,sticky="nsew")
+    widgets["xbarConfigPannel"].grid(columnspan=4, row=7, column=0, rowspan=2, sticky="nsew")# After clicking the tile, the pannel will fill all directions
```
nit: move the comment above the corresponding code

tancheng replied:
I mean for all the comments, they should be above the corresponding code. Please make sure all of them are resolved.

tancheng commented on launchUI.py:
```diff
@@ -696,7 +693,7 @@ def clickSPM():
     spmConfigPannel = widgets["spmConfigPannel"]
     spmConfigPannel.config(text='DataSPM outgoing links')
-    spmConfigPannel.grid(row=7,column=0,rowspan=2,columnspan=4,sticky="nsew")
+    spmConfigPannel.grid(row=7, column=0, rowspan=2, columnspan=4, sticky="nsew") # After clicking the SPM, the pannel will fill all directions
```
ditto

tancheng commented on launchUI.py:
```diff
@@ -1467,7 +1463,7 @@ def create_cgra_pannel(master, rows, columns):
             button = tkinter.Button(canvas, text = "Tile "+str(tile.ID), fg='black', bg='gray', relief='raised', bd=BORDER, command=partial(clickTile, tile.ID))
             posX, posY = tile.getPosXY()
-            canvas.create_window(posX, posY, window=button, height=TILE_HEIGHT, width=TILE_WIDTH, anchor="nw")
+            canvas.create_window(posX, posY, window=button, height=TILE_HEIGHT, width=TILE_WIDTH, anchor="nw") # Tiles will  be placed near the Data memory 
```
nit: two spaces between `will` and `be`.

tancheng commented on launchUI.py:
```diff
@@ -1571,69 +1567,77 @@ def place_xbar_options(master):
+    xbarConfigPannel.columnconfigure(0, weight=1)
+    xbarConfigPannel.columnconfigure(1, weight=1)
+    xbarConfigPannel.columnconfigure(2, weight=1)
+    xbarConfigPannel.columnconfigure(3, weight=1) # Use columnconfigure and rowconfigure to partition the columns, so that each sub-module fills the corresponding space  
```
Where is `rowconfigure`? So there are 4 columns? What are the 4 sub-modules?

Wang-yuanlu replied:
I think I used an inappropriate description. Using "sub-modules" to describe it is not very appropriate. I arranged the 12 function buttons in the xbar window in 3 rows and 4 columns. The code part is to align the 4 divided columns in the xbarpannel so that they can change with the window size.

tancheng reviewed (commented):
Hmmm, which kernel you tried to compile? It also failed on my machine/docker... (I also create `build` folder and typed `python ../launchUI.py`. Or do you mean we need to do `python /WORK_REPO/CGRA-Flow/launchUI.py`?

tancheng commented on launchUI.py:
```diff
@@ -42,8 +39,8 @@ def window_size(window, width, height):
-default_width = 1725
-default_height = 970
+default_width = 1723
+default_height = 900
```
Is this `default_width` and `default_height` for the entire window? If so, can you add comment above it to mention it is for the entire window?

Also, can you please try to decrease the upper part height? For now, the upper part (e.g., CGRA panel) has much redundant space.

Wang-yuanlu commented:
Hi Mr.Tan.I have fine-tuned the initial size of the GUI to reduce redundant space. I adjusted the relative size of several functional modules and moved the tile's positions. For SPM, there were no y-axis coordinates in the initial setup so I chose to extend the height of SPM to accommodate different scales of CGRA. To ensure visual aesthetics, the size of the mapping section is not adjusted. I hope the adjustment is proper.
Have a nice day!

tancheng commented:
Hi Yuanlu, thanks for taking care of my comments.
For the upper part size, I saw you increase the height of the SPM but didn't change the size of tile, which makes them not aligned with each other. Can you revert your changes for this? Let's make them align with each other.

You can see from the figure I attached, can we make the height of the upper part a little bit shorter (instead of just shortening the total window size)? Also, once you make the changes, can you please paste the screenshot here? Thanks.
![Screenshot 2024-06-11 at 7 46 25 AM](https://github.com/tancheng/CGRA-Flow/assets/6756658/febb32e2-5bbb-429a-b8c9-de37430ec289)

Wang-yuanlu commented:
Hi Mr.Tan. I reviewed the materials and adjusted some of the code contents. I divided the GUI interface into two rows. The following is the effect displayed on my computer. The cgrapanel contains additional components such as tiles and SPM. In order to ensure that the GUI has the ability to adapt to different scales of CGRA models, this panel is left with some gaps. I attempted to center these components and reduce the size of the gaps, but the effect was similar to the previous version. It was not very significant.
![GUI](https://github.com/tancheng/CGRA-Flow/assets/80200672/143f291a-c7f4-416a-9eaa-caf67abaf362)
![GUI 2](https://github.com/tancheng/CGRA-Flow/assets/80200672/02f55264-c138-4676-a532-06973cf68663)

tancheng commented on launchUI.py:
```diff
@@ -1568,69 +1565,72 @@ def place_xbar_options(master):
-        paramPannel.columnconfigure(i,weight=1)
+        paramPannel.rowconfigure(i, weight=1)
+    for i in range(3):
+        paramPannel.columnconfigure(i, weight=1) # Use columnconfigure and rowconfigure to partition the columns, so that each column and row will fill the corresponding space
```
Move comment above the code.

tancheng commented on launchUI.py:
```diff
@@ -1568,69 +1565,72 @@ def place_xbar_options(master):
-            paramPannel.rowconfigure(i, weight=2)
-    for i in range(4):
-        paramPannel.columnconfigure(i,weight=1)
+        paramPannel.rowconfigure(i, weight=1)
```
Please comment on `weight`.

Wang-yuanlu replied:
Okay. Here the entire paramPannel is divided into many rows and columns, and the weights are allocated according to the size of the components to be occupied. The “weight” represents the weight of the corresponding row/column length.

tancheng replied:
Ah, i mean add the comment above the corresponding code.

Wang-yuanlu replied:
Okay I will do it.

Wang-yuanlu replied:
There were some issues with my local environment, and a clash occurred, leading to an inability to sync the changes. I will submit the modified version in the future.

tancheng reviewed (commented):
Screenshot looks great to me. Just some nit to be fixed.

tancheng commented:
Sure, please let me do once the environment issue is fixed so we can merge.

tancheng commented on launchUI.py:
```diff
@@ -1566,8 +1566,9 @@ def place_xbar_options(master):
     paramPannel = tkinter.LabelFrame(master, text='Configuration', bd=BORDER, relief='groove')
     paramPannel.grid(row=0, column=1, rowspan=1, columnspan=1, sticky="nsew")
+
     # Use columnconfigure and rowconfigure to partition the columns, so that each column and row will fill the corresponding space
```
nit: period is missing at the end of comment.

Wang-yuanlu replied:
Okay. I will add that period.

Wang-yuanlu commented:
Hi, Mr Tan. I realized the format needs to be consistent so after checking all the added comments, I have unified them into a form without periods. I hope this time the file will look more fluid and aesthetically pleasing so that the subsequent merge and refinement can be more smooth.

tancheng commented:
SG. Is the environment issue fixed on your side and we are ready to merge this PR?

Wang-yuanlu commented:
> SG. Is the environment issue fixed on your side and we are ready to merge this PR?

Yes. Yesterday the lab's server encountered a DNS pollution fault. I modified the host configuration file and now the environment on my side is back to normal.

tancheng commented:
Cool, thanks!
